src/test/java/CompatibilityTest.java

```diff
@@ -31,6 +33,32 @@ public void v2() throws Exception {
         assertFieldType(i, String[].class, "y");
         assertFieldType(i, Object.class, "z");
     }
+    
+    @Test
+    public void testClassesAreOnlyReWrittenWhenNeeded() throws Exception {
+        System.out.println("rewrtting tests");
+        final Transformer t = new Transformer();
+        AntClassLoader cl = new AntClassLoader() {
+            @Override
+            protected Class<?> defineClassFromData(File container, byte[] classData, String classname) throws IOException {
+                byte[] rewritten = t.transform(classname, classData);
+                if (rewritten!=classData ) {
+                    fail(classname + " was rewritten without need");
+                }
+                return super.defineClassFromData(container, rewritten, classname);
+            }
+        };
+
+        cl.addPathComponent(new File("target/test-classes/v2"));
```

Comment:
Or just copy an example of a problematic class file to `src/test/resources/` and load it directly and call `transform` without using a `ClassLoader` at all.